🐛 [Bug] Cross-compiling & -march=native

## 🐛 Bug report
```
uname -a
`Linux ubuntu 5.19.0-26-generic #27-Ubuntu SMP PREEMPT_DYNAMIC Wed Nov 23 20:44:15 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```
```
printenv
SHELL=/bin/bash
PWD=/home/ali/vmangos/build
LOGNAME=ali
XDG_SESSION_TYPE=tty
CXX=/usr/bin/aarch64-linux-gnu-g++
CXXFLAGS=-march=armv8-a+crc -mtune=cortex-a72
TBB_ROOT_DIR=/usr/include/tbb
HOME=/home/ali
LANG=en_US.UTF-8
ACE_ROOT=/usr/include/ace
SSH_CONNECTION=192.168.1.2 60229 192.168.1.10 22
LESSCLOSE=/usr/bin/lesspipe %s %s
XDG_SESSION_CLASS=user
TERM=xterm-256color
LESSOPEN=| /usr/bin/lesspipe %s
USER=ali
SHLVL=1
XDG_SESSION_ID=1
XDG_RUNTIME_DIR=/run/user/1000
SSH_CLIENT=192.168.1.2 60229 22
XDG_DATA_DIRS=/usr/local/share:/usr/share:/var/lib/snapd/desktop
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
SSH_TTY=/dev/pts/0
OLDPWD=/home/ali/vmangos/core
_=/usr/bin/printenv
```

```
cmake ~/vmangos/core -DDEBUG=0 -DSUPPORTED_CLIENT_BUILD=5875 -DUSE_EXTRACTORS=0 -DCMAKE_INSTALL_PREFIX=~/vmangos
-- Detected 64-bit platform.
--
This script builds the MaNGOS server.
  Options that can be used in order to configure the process:
    PREFIX: Path where the server should be installed to
    USE_PCH: Use precompiled headers
  To set an option simply type -D<OPTION>=<VALUE> after 'cmake <srcs>'.
  For example: cmake .. -DPREFIX=/opt/mangos

-- Found ACE library: /usr/lib/aarch64-linux-gnu/libACE.so
-- Found ACE headers: /usr/include
-- Using mysql-config: /usr/bin/mysql_config
-- Found MySQL library: /usr/lib/aarch64-linux-gnu/libmysqlclient.so
-- Found MySQL headers: /usr/include/mysql
--
-- Core revision: d58ade9a73ea090e56d0 2022-12-31 13:38:40 +0200
-- Supported version: 5875
-- Install server to: /home/ali/vmangos
--
-- Use PCH               : Yes
-- PGO profile           : Disabled
-- Memory allocation     : TBB/Release /usr/lib/aarch64-linux-gnu/libtbbmalloc.so;/usr/lib/aarch64-linux-gnu/libtbb.so
-- Detected compiler     : GNU
-- Build type            : Release
-- Use anticheat         : No  (default)
-- Build scripts         : Yes (default)
-- Debug symbols         : Included
-- Build flags (RELEASE) : -O3 -DNDEBUG  -g --no-warnings -fexceptions -fnon-call-exceptions -march=native -pipe
-- Build flags (DEBUG)   : -g  -g --no-warnings -fexceptions -fnon-call-exceptions -march=native -pipe
-- Configuring done
-- Generating done
-- Build files have been written to: /home/ali/vmangos/build
```
```
make
[  0%] Building CXX object src/shared/CMakeFiles/shared.dir/Common.cpp.o
Assembler messages:
Error: unknown architecture `native'

Error: unrecognized option -march=native
cc1plus: error: unknown value ‘native’ for ‘-march’
cc1plus: note: valid arguments are: armv8-a armv8.1-a armv8.2-a armv8.3-a armv8.4-a armv8.5-a armv8.6-a armv8.7-a armv8.8-a armv8-r armv9-a
make[2]: *** [src/shared/CMakeFiles/shared.dir/build.make:76: src/shared/CMakeFiles/shared.dir/Common.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:320: src/shared/CMakeFiles/shared.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```
Problem can be fixed by removing `-macrh=native` option from CMakeLists.txt file line 356.

[CMakeCache.txt](https://github.com/vmangos/core/files/10327841/CMakeCache.txt)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -32,6 +32,7 @@ set(CMAKE_MODULE_PATH
 
 set(SUPPORTED_CLIENT_BUILD "CLIENT_BUILD_1_12_1" CACHE STRING "Client version the core will support")
 option(USE_STD_MALLOC "Use standard malloc instead of TBB" OFF)
+option(BUILD_FOR_HOST_CPU "Build specifically for the host CPU via `-march=native` (might not run on different machines)" ON)
 option(TBB_DEBUG "Use TBB debug librairies" OFF)
 option(USE_ANTICHEAT "Use anticheat" OFF)
 option(USE_SCRIPTS "Compile scripts" ON)
@@ -78,7 +79,7 @@ if(WIN32)
       message(FATAL_ERROR "Only Visual Studio 2015 or newer is supported")
     endif()
   endif()
-  
+
   # Added by Giperion, in WIN32 output all compiled files in one directory
   set( CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_SOURCE_DIR}/bin )
   set( CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_SOURCE_DIR}/lib )
@@ -196,13 +197,13 @@ if(NOT USE_STD_MALLOC)
       ${TBB_LIBRARIES}
     )
   endif()
-  
+
   if (WIN32)
     list (GET TBB_LIBRARIES 1 TBB_SINGLE_FILEDIRECTORY)
     get_filename_component(TBB_LIB_DIRECTORY "${TBB_SINGLE_FILEDIRECTORY}" DIRECTORY)
     link_directories(${TBB_LIB_DIRECTORY})
   endif()
-  
+
 endif()
 
 # Win32 delifered packages
@@ -254,7 +255,7 @@ else()
     OUTPUT_STRIP_TRAILING_WHITESPACE
     ERROR_QUIET
   )
-  
+
   if(NOT rev_hash)
     # No valid ways available to find/set the revision/hash, so let's force some defaults
     message(STATUS "
@@ -263,7 +264,7 @@ else()
     set(rev_date "1970-01-01 00:00:00 +0000")
     set(rev_hash "unknown")
   endif()
-  
+
   configure_file(${CMAKE_CURRENT_SOURCE_DIR}/cmake/revision.h.cmake ${CMAKE_CURRENT_SOURCE_DIR}/src/shared/revision.h)
 endif()
 
@@ -316,6 +317,12 @@ endif()
 message(STATUS "Memory allocation     : ${ALLOC_LIB_INFO_STRING} ${TBB_LIBRARIES}")
 message(STATUS "Detected compiler     : ${CMAKE_CXX_COMPILER_ID}")
 
+if(BUILD_FOR_HOST_CPU)
+  message(STATUS "Build for host CPU    : Yes (default)")
+else()
+  message(STATUS "Build for host CPU    : No")
+endif()
+
 message(STATUS "Build type            : ${CMAKE_BUILD_TYPE}")
 
 if(USE_ANTICHEAT)
@@ -338,7 +345,7 @@ if(UNIX)
   else()
     message(STATUS "Debug symbols         : Disabled")
   endif()
-  
+
   if(GCC_SANITIZE)
     set(BUILD_ADDITIONAL_FLAGS "${BUILD_ADDITIONAL_FLAGS} -fno-omit-frame-pointer -fno-optimize-sibling-calls")
     set(BUILD_ADDITIONAL_FLAGS "${BUILD_ADDITIONAL_FLAGS} -fsanitize=address")
@@ -353,7 +360,10 @@ if(UNIX)
     #set(BUILD_ADDITIONAL_FLAGS "${BUILD_ADDITIONAL_FLAGS} -fsanitize=unsigned-integer-overflow")
   endif()
 
-  set(BUILD_ADDITIONAL_FLAGS "${BUILD_ADDITIONAL_FLAGS} --no-warnings -fexceptions -fnon-call-exceptions -march=native -pipe ")
+  set(BUILD_ADDITIONAL_FLAGS "${BUILD_ADDITIONAL_FLAGS} --no-warnings -fexceptions -fnon-call-exceptions -pipe")
+  if(BUILD_FOR_HOST_CPU)
+    set(BUILD_ADDITIONAL_FLAGS "${BUILD_ADDITIONAL_FLAGS} -march=native")
+  endif()
   set(CMAKE_CXX_STANDARD 14)
   if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
     if(PROFILE_GENERATE)
@@ -392,7 +402,7 @@ if(MSVC)
   #if(PLATFORM MATCHES X86)
   #  set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} /LARGEADDRESSAWARE")
   #endif()
-  
+
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /bigobj")
 endif()
 

